### ISSUE TITLE
Disambiguated (Label-based) graph's 'compact' flag not encompassing all use cases

### ISSUE THREAD

**Issue by chrabyrd** [archesproject/arches#7139]
## This ticket is meant to capture discussion among Farallon developers.

Currently accessing the `resources/${resource_id}` API returns a disambiguated graph with the `compact` flag set to `true` by default. However, there is an edge case in arches-HER that breaks the core logic of the `compact` flag , and it's common enough to harden our core logic against.

## Consider the following example

UNCOMPACTED graph: 

    "displaydescription": "<Description>",
    "displayname": "sdfg",
    "graph_id": "076f9381-7b00-11e9-8d6b-80000b44d1d9",
    "legacyid": null,
    "map_popup": " sdfg",
    "resource": {
        "Heritage Asset Names": {
            "@node_id": "676d47f9-9c1c-11ea-9aa0-f875a44e0e11",
            "@tile_id": "c531f990-ab5f-4ef2-b83f-177e5f9c213a",
            "@value": "NON_DATA_COLLECTING_NODE",
            "Asset Name": {
                "@node_id": "676d47ff-9c1c-11ea-b07f-f875a44e0e11",
                "@tile_id": "c531f990-ab5f-4ef2-b83f-177e5f9c213a",
                "@value": "sdfg"
            },
            "Asset Name Currency": {
                "@node_id": "676d47fd-9c1c-11ea-9d73-f875a44e0e11",
                "@tile_id": "c531f990-ab5f-4ef2-b83f-177e5f9c213a",
                "@value": "Former",
                "Asset Name Currency Metatype": {
                    "@node_id": "676d4802-9c1c-11ea-a30f-f875a44e0e11",
                    "@tile_id": "c531f990-ab5f-4ef2-b83f-177e5f9c213a",
                    "@value": "Currency Type"
                }
            },
        }
    },
    "resourceinstanceid": "e954c3ca-64de-411a-8d92-f1a45258e7d2"

COMPACTED graph: 

    "displaydescription": "<Description>",
    "displayname": "sdfg",
    "graph_id": "076f9381-7b00-11e9-8d6b-80000b44d1d9",
    "legacyid": null,
    "map_popup": " sdfg",
    "resource": {
        "Heritage Asset Names": {
            "Asset Name": "sdfg",
            "Asset Name Currency": {
                "Asset Name Currency Metatype": "Currency Type"
            },
        }
    },
    "resourceinstanceid": "e954c3ca-64de-411a-8d92-f1a45258e7d2"

We can see the `compact` logic working as intended with `Heritage Asset Names` and `Asset Name`. 
- `Heritage Asset Names` has children, but does not collect data
- `Asset Names` collects data but has no children.

## Paths forward

There are a three main potential paths forward here. The purpose of this discussion is to select one, or find a better path to suit our needs.

#### Edit the `compact` flag to decouple value from children
- If we denormalize the graph then will not break current API
- We already do this with the uncompacted graph. 
- Can keep `@value` or use `*__value` key
- Should this structure be applied throughout, or only on nodes that meet a certain criteria?
- Does the break the 'spirit' of a `compact` flag? Is there that much difference between compacted and uncompacted?

#### Remove the `compact` flag
- Straightforward to remove
- Reduces complexity / allows for frontend formatting
- Breaks any system using current shape ( Dashboards )

#### Keep the borked `compact` logic && current API, to be deprecated later in favor of a non-compacted API
- Small overhead in building 2nd API
- Only move systems to using the new API if they need to currently
- Backwards-compatible && friendly to external devs
- Make a well-promulgated announcement about deprecation

## How to help

Let's have a discussion as a core team! What are your thoughts and feelings about the current situation, and what do you see ( listed or unlisted ) as the best potential path forward?

**Comment by chiatt** [comment]
Wouldn't the first option also break existing Dashboards?

**Comment by chrabyrd** [comment]
> Wouldn't the first option also break existing Dashboards?

Potentially, depending on implementation. If we decide to break graph normalization then no, for instance

**Comment by apeters** [comment]
I'm leaning toward continuing to support the "compact" flag and having it be as compact as possible.  That means only using the @value property where needed.  IE: only were the root node of a card collects values as in the example below.

```
"displaydescription": "<Description>",
"displayname": "sdfg",
"graph_id": "076f9381-7b00-11e9-8d6b-80000b44d1d9",
"legacyid": null,
"map_popup": " sdfg",
"resource": {
    "Heritage Asset Names": {
        "Asset Name": "sdfg", 
        "Asset Name Currency": {
            "@value": "Former",
            "Asset Name Currency Metatype": "Currency Type"
        },
    }
},
"resourceinstanceid": "e954c3ca-64de-411a-8d92-f1a45258e7d2"
```

**Comment by chrabyrd** [comment]
> I'm leaning toward continuing to support the "compact" flag and having it be as compact as possible. That means only using the @value property where needed. IE: only were the root node of a card collects values as in the example below.
> 
> ```
> "displaydescription": "<Description>",
> "displayname": "sdfg",
> "graph_id": "076f9381-7b00-11e9-8d6b-80000b44d1d9",
> "legacyid": null,
> "map_popup": " sdfg",
> "resource": {
>     "Heritage Asset Names": {
>         "Asset Name": "sdfg", 
>         "Asset Name Currency": {
>             "@value": "Former",
>             "Asset Name Currency Metatype": "Currency Type"
>         },
>     }
> },
> "resourceinstanceid": "e954c3ca-64de-411a-8d92-f1a45258e7d2"
> ```

Thinking programatically, does that mean that every system that consumes this API needs explicit instruction && supporting logic around the check for a `@value` sub-node as default code ?

If de-normalizing the graph is the path forward, then I say we should make the documented default public API have `compacted=false` . This doesn't necessarily mean changing the current API this second, but ( in my mind ) rather gracefully deprecating for another.

**Comment by chiatt** [comment]
I see the value in minimizing the payload of a request by making the response as compact as possible. However, (as @chrabyrd points out) if we only use @value where it’s absolutely needed, it does put the burden on the developer to understand the schema of their model or at least understand why @value exists in some cases and why it doesn’t in others. Using ‘@value’ for every node value seems like a reasonable compromise being moderately compact and having more intuitive interface.

There may be other reasons for keeping the highly compacted endpoint such as human readability and the fact that existing systems rely on it. However, in future Arches versions, I would be in favor of making the default response either non-compacted or moderately compacted (‘@value’ on every node). We can then decide to fix or deprecate the highly compacted endpoint.

**Comment by apeters** [comment]
I would vigorously argue that anyone using this output will most definitely have to understand the graph shape to be able to understand how to parse the JSON, this goes for the un-compacted version as well.  I think introducing a property that will ALWAYS be null and can never have a value is just misleading at worst and annoying at best.  I think we're getting hung up on @value and focusing on it as if it's special; it's not.  Think of it like you would any other property.  

Do you question why the property "Asset Name Currency Value" is in the following snippet?  No, you just accept that that is the shape of the JSON.

```
    "Heritage Asset Names": {
        "Asset Name": "Taco",
        "Asset Name Currency": {
            "Asset Name Currency Value": "Former",
            "Asset Name Currency Metatype": "Currency Type"
        }
    }
```

Now, what if the shape was this (Does this change your thinking?):
```
    "Heritage Asset Names": {
        "Asset Name": "Taco",
        "Asset Name Currency": {
            "Value": "Former",
            "Asset Name Currency Metatype": "Currency Type"
         }   
     }
```

Now, what about this?:
```
    "Heritage Asset Names": {
        "Asset Name": "Taco",
        "Asset Name Currency": {
            "@value": "Former",
            "Asset Name Currency Metatype": "Currency Type"
        }   
    }
```

What is being suggested is something like this:
```
    "Heritage Asset Names": {
        "@value": "NON_DATA_COLLECTING_NODE",
        "Asset Name": "Taco",
        "Asset Name Currency": {
            "@value": "Former",
            "Asset Name Currency Metatype": "Currency Type"
        }   
    }
```

Magic strings are a pattern that we should avoid if possible.  It just adds cruft and really adds no value.  
@chrabyrd, To your point about systems that would consume this api --> As a developer of said system, they would have to understand that they would need to parse that node("@value": "NON_DATA_COLLECTING_NODE") and ignore it.  

Here's some example parsing logic assuming the solution suggested by @chiatt and @chrabyrd:
```
if object.@value !== "NON_DATA_COLLECTING_NODE":
   do something...
```
OR with the solution I'm suggesting
```
if object.@value !== undefined:
   do something
```
Either way a consumer of the api will have to have some logic to account for those properties.
Hopefully someone doesn't save a node with the actual value of "NON_DATA_COLLECTING_NODE".  That's why magic strings are something to be avoided if possible.  

In the scenario I'm proposing, @value will ALWAYS have a valid value.

The more I think about it, I would suggest the same thing for our non-compacted version (but that's maybe a discussion for a different day).

**Comment by chrabyrd** [comment]
```
 "Heritage Asset Names": {
        "@value": "NON_DATA_COLLECTING_NODE",
        "Asset Name": "Taco",
        "Asset Name Currency": {
            "@value": "Former",
            "Asset Name Currency Metatype": "Currency Type"
        }   
    }
```

Not quite, I'm envisioning no graph de-normalization with the end user needing to know very little about the api. Easy to replace `NON_DATA_COLLECTING_NODE` in the backend. In my mind this example looks like:

```
 "Heritage Asset Names": {
        "@value": "null",
        "Asset Name": {
            "@value": "Taco",
         },
        "Asset Name Currency": {
            "@value": "Former",
            "Asset Name Currency Metatype": {
                "@value": Currency Type"
             }
        }   
    }
```


In the above the graph is normalized and the API consumer doesn't need to know any special behind-the-scenes logic. She/He/It knows only once source of truth for value, the `@value` node.



In my mind, this is the correct path to take. I have a strong opinion that unnecessarily denormalizing the graph is not the best move we could make. That said, I _do_ respect organizational pecking order and I feel that I've done my part in making my professional opinion heard. @apeters if you believe that we are truly crosspoint on this I have no qualms enacting your solution. Please let me know your thoughts an opinions when you're able, I should be rolling onto this this week.

**Comment by apeters** [comment]
I think we're getting hung up on the @ symbol and that it somehow denotes something special.  I really doesn't.  Forcing the @value property to be used everywhere there is a literal is completely unnecessary and in some cases actually misleading.  There is no need for that.  I actually don't understand what a "normalized" json object is.  A normalized database yes.  

See the example tweet object below.  Is it normalized?  I have no idea, but what I do know is that to consume data from Twitter, I will have to look at the docs and understand the shape of the json object returned from calls to the API.  

The argument that because every leaf node ends in "@value" somehow removes that burden is false.  I would still have to consult the API.  Not only that but I would have to understand that in some instances the "@value" property is just an empty container that will never hold a true value.

Example Tweet Object from Twitter
```
{
  "created_at": "Thu Apr 06 15:24:15 +0000 2017",
  "id_str": "850006245121695744",
  "text": "1\/ Today we\u2019re sharing our vision for the future of the Twitter API platform!\nhttps:\/\/t.co\/XweGngmxlP",
  "user": {
    "id": 2244994945,
    "name": "Twitter Dev",
    "screen_name": "TwitterDev",
    "location": "Internet",
    "url": "https:\/\/dev.twitter.com\/",
    "description": "Your official source for Twitter Platform news, updates & events. Need technical help? Visit https:\/\/twittercommunity.com\/ \u2328\ufe0f #TapIntoTwitter"
  },
  "place": {   
  },
  "entities": {
    "hashtags": [      
    ],
    "urls": [
      {
        "url": "https:\/\/t.co\/XweGngmxlP",
        "unwound": {
          "url": "https:\/\/cards.twitter.com\/cards\/18ce53wgo4h\/3xo1c",
          "title": "Building the Future of the Twitter API Platform"
        }
      }
    ],
    "user_mentions": [     
    ]
  }
}
```

Maybe the idea that having "@value" used for every literal means that it eases logic to automatically consume a json object.  Once again, I think that's false.  JSON traversal logic does not rely on hardcoded property names.

In the above Tweet object one doesn't question why the "urls" and "unwound" object both have a "url" property but the root object doesn't.  You just accept that the json object has that shape because the developers designed it that way.  If it were "@url" instead it wouldn't change a thing.  These objects aren't "normalized".  Each one defines the qualities of a specific entity and no more.

In the end, my suggestion is to keep the "compact" flag on our API and only use the "@value" where we need to.  This should fix the issue with the current code with the least amount of disruption to existing users.

**Comment by chrabyrd** [comment]
I feel like we're arguing different points.

To clarify my nomenclature, I'm using 'normalized' in the context of an API response. I'm using it to mean that an API consumer only needs to look in one place for a wanted value; they would not need to program `if` logic into the functions that read over the data returned, because the graph shape doesn't change.

I'm not arguing that an API consumer doesn't need to know _anything_ about the API it's consuming.

**In the Twitter example provided, I'd argue the response is 'normalized'.** 

The graph shape doesn't change if a tweet has, say, 0 hashtags, 1 hashtag, 2 hashtags, or 50. The API consumer knows exactly where to look for hashtags.

Or take the `urls` example in the tweet object above. It's an array of objects, an each object has `url` and `unwound` keys. The `unwound` object has `url` and `title` keys. I haven't dug into the API docs but I'd bet money this shape never changes, regardless of data present. 

Again, I haven't dug too deeply into the Twitter API but because there's empty values present in the example response I'd bet the tweet graph doesn't change shape. Each datum lives in a single place; behind a singe 'path' so-to-say.

Having unchanging logic for when we use a `@value` key gives us this same unchanging graph shape. 

The API consumer wouldn't need to build logic to first look in one place for a value ( the `Asset Name` example ), but if it's not there look in another place ( the `Asset Name Currency`[`@value`] example ). They'd just need to look in one place `thing`[`@value`].


Again this is to clarify/defend my opinion. I have no qualms enacting a solution I don't agree with.

**Comment by chrabyrd** [comment]
@apeters 

Currently using this space to talk about a **COMPLETELY SEPARATE** label_based_graph issue. I'm not trying to raise the dead :-)

I think I was conflating 2 issues in my arguments against the `compact` changes. Below I'll layout the conflating argument which still potentially holds.

This has nothing to do with `compact`, I'm proposing potentially altering the return of core response of label_based_graph.

This issue has fallen out as I've been developing against the uncompacted label_based_graph API. This is fortunate because it provides a reproducible real-world example.

Let's say I'm in AfS and I've created a `Physical Thing` with a single name.

ex snippet:

    "resource": {
        "Name": {
            "Name_content": "uf7",
            "Name_label ": null,
            "Name_language ": "English (language)",
            "Name_source ": "",
            "Name_type ": "preferred terms"
        },


Now let's say I've created a `Physical Thing` with two names.

ex snippet:

    "resource": {
        "Name": [
            {
                "Name_content": "viucivuuvi",
                "Name_label ": null,
                "Name_language ": "English (language),Danish (language)",
                "Name_source ": "",
                "Name_type ": "preferred terms,informal title"
            },
            {
                "Name_content": "afw33f3f3f3f3f3",
                "Name_label ": null,
                "Name_language ": "Chinese (language)",
                "Name_source ": "",
                "Name_type ": "informal title"
            }
        ]
    },



As a consumer of this API, there is currently no way of knowing if it's possible for a `Physical Thing` to have one or many names. Therefore when building out the consuming code, I need to build in some complexity to gracefully handle for if a `Physical Thing` has one or many names, and is therefore an array or an Object.

ex snippet:

                var resourceInstanceData = Object.values(disambiguatedResourceData).find(function(topLevelData) {
                    if (topLevelData instanceof Array) {
                        return topLevelData[0][NODE_ID] === section.nodeId;  /* all data has same nodeId */
                    }
                    return topLevelData[NODE_ID] === section.nodeId;
                });

                if (!(resourceInstanceData instanceof Array)) {
                    resourceInstanceData = [ resourceInstanceData ];
                } 

The above is client-side code that coerces all top-level data into arrays, so that the following logic only needs to handle a single data type.

This is something **every** API consumer will need to handle for. API consumers have no way of determining whether a node value will be an array or an object. We currently have no flag for cardinality. It then seems very likely that most API consumers will have some sort of coercing code on their end to do similar to the above.

This is the denormalization I was trying to explain. The same node for 2 instances of the same resource can be an array or an object; there's no way to tell what shape it will take until we read it. It forces the normalization code ( eg coercing into arrays ) to exist in the API consumer's ecosystem.

My argument is we need to handle for cardinality-n in this API. If a node value has the potential to be an array, it should _always_ be an array. To have it otherwise denormalizes the graph shape and forces the API consumer to have code that handles this. 

This is a generic/common enough issue with our graph shape that I feel we should take the burden of coercing datatypes away from the consumer. I fear that all API consumers across the board will need some sort of code to coerce these values into a normalized shape, and we could easily remove that burden by handling it ourselves within the endpoint.
